### PR TITLE
config: add warning, when users set configuration to mydumper/loader/syncer b…

### PR DESCRIPTION
**Pull request by zeminzhou**
…ut not use them

### What problem does this PR solve? 
For some configurations like routes/filters/mydumpers/loaders/syncers that users set in global configuration, this pr add “reference count” for these configurations.

### What is changed and how it works?
When some configurations are set in global configuration, but instances don't use thme. An error will return.

### Tests
+ unit test

### CODE REVIEW

**Review by lichunzhu — COMMENTED** [review]
I don't think users will check dm-worker's log when they start a task. I think we should print these messages to `result.Msg` or something else.

**Comment by zeminzhou** [comment]
> I don't think users will check dm-worker's log when they start a task. I think we should print these messages to `result.Msg` or something else.

good idea

File: dm/config/task.go [hunk 1]
```diff
@@ -418,6 +418,8 @@ func (c *TaskConfig) adjust() error {
 	}
 
 	iids := make(map[string]int) // source-id -> instance-index
+	gobalConfigReferCount := map[string]int{}
```
Comment (lance6716) [inline dm/config/task.go:421]:
```suggestion
	globalConfigReferCount := map[string]int{}
```

File: dm/config/task.go [hunk 2]
```diff
@@ -472,10 +477,15 @@ func (c *TaskConfig) adjust() error {
...
 				return terror.ErrConfigMydumperCfgNotFound.Generate(i, inst.MydumperConfigName)
 			}
+			gobalConfigReferCount[prefixs[3]+inst.MydumperConfigName]++
 			inst.Mydumper = new(MydumperConfig)
 			*inst.Mydumper = *rule // ref mydumper config
 		}
 		if inst.Mydumper == nil {
+			if len(c.Mydumpers) != 0 && inst.MydumperThread == 0 {
+				log.L().Warn("mysql instance don't refer mydumper's configuration with mydumper-config-name, the default configuration will be used", zap.Int("mysql instance", i))
+				return terror.ErrConfigUnreferingCofig.Generate()
```
Comment (lance6716) [inline dm/config/task.go:487]: If we let DM fallback to default config, we should not return here

and I think check `inst.MydumperThread == 0` is not necessary

File: dm/config/task.go [hunk 3]
```diff
@@ -539,6 +559,42 @@ func (c *TaskConfig) adjust() error {
 		return terror.ErrConfigDuplicateCfgItem.Generate(strings.Join(duplicateErrorStrings, "\n"))
 	}
 
+	for route := range c.Routes {
+		if gobalConfigReferCount[prefixs[0]+route] == 0 {
+			log.L().Error("route in gobal configuration isn't used", zap.String("route", route))
+			return terror.ErrConfigGobalConfigUnused.Generate()
```
Comment (lance6716) [inline dm/config/task.go:565]: we should return the unused configs in error message, grep the names in log is a hard work.

could add the names in a slice and format error message with all of them.

File: dm/config/task_test.go [hunk 4]
```diff
@@ -28,6 +28,314 @@ import (
...
+    syncer-config-name: "global1"
+
+  - source-id: "mysql-replica-02"
+    route-rules: ["route-rule-1"]
+    filter-rules: ["filter-rule-1"]
+    column-mapping-rules: ["column-mapping-rule-1"]
+`
+	taskConfig = NewTaskConfig()
+	err = taskConfig.Decode(errorTaskConfig2)
+	c.Assert(err, NotNil)
```
Comment (lance6716) [inline dm/config/task_test.go:336]: after we return the unused names, we should use `ErrorMatches` to check them

File: pkg/terror/error_list.go [hunk 5]
```diff
@@ -823,6 +824,7 @@ var (
 	ErrConfigEtcdParse              = New(codeConfigEtcdParse, ClassConfig, ScopeInternal, LevelHigh, "incapable config of %s from etcd", "")
 	ErrConfigMissingForBound        = New(codeConfigMissingForBound, ClassConfig, ScopeInternal, LevelHigh, "source bound %s doesn't have related source config in etcd", "")
 	ErrConfigBinlogEventFilter      = New(codeConfigBinlogEventFilter, ClassConfig, ScopeInternal, LevelHigh, "generate binlog event filter", "Please check the `filters` config in source and task configuration files.")
+	ErrConfigGlobalConfigsUnused    = New(codeConfigGlobalConfigsUnused, ClassConfig, ScopeInternal, LevelHigh, "The configurations as following %v are set in gobal configuration but instances don't use them", "Please check the configuration files.")
```
Comment (lance6716) [inline pkg/terror/error_list.go:827]:
```suggestion
	ErrConfigGlobalConfigsUnused    = New(codeConfigGlobalConfigsUnused, ClassConfig, ScopeInternal, LevelHigh, "The configurations as following %v are set in global configuration but instances don't use them", "Please check the configuration files.")
```

File: dm/config/task.go [hunk 6]
```diff
@@ -512,10 +525,14 @@ func (c *TaskConfig) adjust() error {
 			if !ok {
 				return terror.ErrConfigSyncerCfgNotFound.Generate(i, inst.SyncerConfigName)
 			}
+			globalConfigReferCount[prefixs[5]+inst.SyncerConfigName]++
 			inst.Syncer = new(SyncerConfig)
 			*inst.Syncer = *rule // ref syncer config
 		}
 		if inst.Syncer == nil {
+			if len(c.Syncers) != 0 {
+				log.L().Warn("mysql instance don't refer syncer's configuration with syncer-config-name, the default configuration will be used", zap.Int("mysql instance", i))
```
Comment (lance6716) [inline dm/config/task.go:534]: we could use `source-id` to identify it

**Review by lance6716 — COMMENTED** [review]
rest LGTM

**Comment by lance6716** [comment]
for failed tests, you could sort the slice to get the stable result. rest LGTM

**Review by lichunzhu — COMMENTED** [review]
/lgtm

File: dm/config/task_test.go [hunk 7]
```diff
@@ -234,7 +234,8 @@ mysql-instances:
 	taskConfig = NewTaskConfig()
 	err = taskConfig.Decode(errorTaskConfig)
 	c.Check(err, NotNil)
-	c.Assert(err, ErrorMatches, "*The configurations as following [route-rule-2 filter-rule-2 column-mapping-rule-2] are set in global configuration but*")
+	//c.Assert(err, ErrorMatches, "*The configurations as following [route-rule-2 filter-rule-2 column-mapping-rule-2] are set in global configuration but*")
```
Comment (lance6716) [inline dm/config/task_test.go:237]: this test can't detect only one of reference check not working. `sort.Strings` is a better way

and please remove the commented line above to keep codebase clean

**Review by lichunzhu — COMMENTED** [review]
LGTM

**Review by lichunzhu — APPROVED** [review]
LGTM

**Pull request merged**